### PR TITLE
[Enhancement] improve map columns's remove dup key performence

### PR DESCRIPTION
**Pull request by luohaha**
Improve map columns's remove dup key performence, complexity from O(n^2) to O(n).

## What type of PR is this:
- [ ] BugFix
- [ ] Feature
- [x] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

## Checklist:
- [x] I have added test cases for my bug fix or my new feature
- [ ] This pr will affect users' behaviors
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function

## Bugfix cherry-pick branch check:
- [x] I have checked the version labels which the pr will be auto-backported to the target branch
  - [x] 3.1
  - [x] 3.0
  - [x] 2.5
  - [ ] 2.4

### CODE REVIEW

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]


:heart_eyes: **pass** : 0 / 0 (0%)

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]


:heart_eyes: **pass** : 0 / 0 (0%)

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]


:heart_eyes: **pass** : 0 / 0 (0%)

File: be/src/column/map_column.cpp [hunk 1]
```diff
@@ -714,15 +714,23 @@ void MapColumn::remove_duplicated_keys(bool need_recursive) {
...
+        key_hash_to_offsets.reserve(_offsets->get_data()[i + 1] - _offsets->get_data()[i]);
         for (auto j = _offsets->get_data()[i]; j < _offsets->get_data()[i + 1]; ++j) {
-            for (auto k = j + 1; k < _offsets->get_data()[i + 1]; ++k) {
-                if (hash[j] == hash[k] && _keys->equals(j, *_keys, k)) {
-                    filter[j] = 0;
-                    has_duplicated_keys = true;
-                    break;
+            auto& same_hash_offsets = key_hash_to_offsets[hash[j]];
+            if (!same_hash_offsets.empty()) {
+                for (auto k = 0; k < same_hash_offsets.size(); k++) {
```
Comment (imay) [inline be/src/column/map_column.cpp:722]:
```suggestion
                for (auto offset : same_hash_offsets) {
```

File: be/src/column/map_column.cpp [hunk 2]
```diff
@@ -714,15 +714,23 @@ void MapColumn::remove_duplicated_keys(bool need_recursive) {
 
     uint32_t new_offset = 0;
     for (auto i = 0; i < size; ++i) {
+        std::unordered_map<uint32_t, std::vector<uint32_t>> key_hash_to_offsets;
```
Comment (imay) [inline be/src/column/map_column.cpp:717]: https://en.cppreference.com/w/cpp/container/unordered_multimap
You can use std unordered_multimap directly.

Comment (Seaven) [reply be/src/column/map_column.cpp:717]: set is enough?

Comment (luohaha) [reply be/src/column/map_column.cpp]: Done.

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]


:heart_eyes: **pass** : 0 / 0 (0%)

File: be/test/exprs/lambda_map_expr_test.cpp [hunk 3]
```diff
@@ -251,7 +251,7 @@ TEST_F(MapApplyExprTest, test_map_int_int) {
         ColumnPtr result = map_apply_expr->evaluate(&exprContext, &cur_chunk);
 
         EXPECT_TRUE(result->is_nullable());
-        EXPECT_STREQ(result->debug_string().c_str(), "[{0:67}, {0:89}, {0:NULL}, {}, NULL]");
+        EXPECT_STREQ(result->debug_string().c_str(), "[{0:45}, {0:78}, {0:NULL}, {}, NULL]");
```
Comment (fzhedu) [inline be/test/exprs/lambda_map_expr_test.cpp:254]: why change these tests? the function should keep the last win

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]


:heart_eyes: **pass** : 0 / 0 (0%)

File: be/test/exprs/lambda_map_expr_test.cpp (hunk 3 above)
Comment (imay) [reply be/test/exprs/lambda_map_expr_test.cpp:254]: There is no order in the map definition. So I don't think keeping the last win is necessary.

Comment (fzhedu) [reply be/test/exprs/lambda_map_expr_test.cpp:254]: trino keep the last win

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]


:heart_eyes: **pass** : 0 / 0 (0%)

File: be/test/exprs/lambda_map_expr_test.cpp (hunk 3 above)
Comment (luohaha) [reply be/test/exprs/lambda_map_expr_test.cpp:254]: Ok, I will change this to last win

Comment (luohaha) [reply be/test/exprs/lambda_map_expr_test.cpp]: Done

**Review by fzhedu — APPROVED** [review]

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]


:heart_eyes: **pass** : 0 / 0 (0%)

**Review by Seaven — APPROVED** [review]

**Comment by wanpengfei-git** [comment]
@Mergifyio backport branch-3.1

**Comment by wanpengfei-git** [comment]
@Mergifyio backport branch-3.0

**Pull request merged**